Source/Core/VideoCommon/ShaderGenCommon.cpp

```diff
@@ -362,3 +363,89 @@ const char* GetInterpolationQualifier(bool msaa, bool ssaa, bool in_glsl_interfa
       return "sample";
   }
 }
+
+void WriteCustomShaderStructDef(ShaderCode* out, u32 numtexgens)
+{
+  // Roll this when there are breaking changes to the API
+  out->Write("const uint CUSTOM_SHADER_API_VERSION = 1;\n");
+
+  // CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE "enum" values
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_NONE = {};\n",
+             static_cast<u32>(AttenuationFunc::None));
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_POINT = {};\n",
+             static_cast<u32>(AttenuationFunc::Spec));
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_DIR = {};\n",
+             static_cast<u32>(AttenuationFunc::Dir));
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_ATTENUATION_TYPE_SPOT = {};\n",
+             static_cast<u32>(AttenuationFunc::Spot));
+
+  // CUSTOM_SHADER_LIGHTING_LIGHT_TYPE "enum" values
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_LIGHT_TYPE_ALPHA = 0;\n");
+  out->Write("const uint CUSTOM_SHADER_LIGHTING_LIGHT_TYPE_COLOR = 1;\n");
+
+  out->Write("struct CustomShaderLightData\n");
+  out->Write("{{\n");
+  out->Write("\tfloat3 position;\n");
+  out->Write("\tfloat3 direction;\n");
+  out->Write("\tfloat3 color;\n");
+  out->Write("\tuint attenuation_type;\n");
+  out->Write("\tuint light_type;\n");
```

Comment:
Yeah it might not be.  At the moment, most of my light logic just iterates through the color lights.  I haven't done much investigation into the alpha lights.

On the (very large) todo list is a light visualizer.  I imagine once I have that, it'll be easier to understand what is going on with those four light categories.